From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Mon, 2 Nov 2020 00:09:25 -0800
Subject: [PATCH] Filter useless entity packets

This patch prevents useless packets (zero movement packets) from being sent in the entity tracker. This is especially useful with the async entity tracker optimization since it prevents unnecessary runnables in the processQueue, overall improving performance.

diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 0a94e04205c1f57a5956ea5945fdff543667e58a..685e5d7738d83e0642b4e0601a6abf668dc85428 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -162,11 +162,14 @@ public class EntityTrackerEntry {
                     if (!flag4 && this.o <= 400 && !this.q && this.r == this.tracker.isOnGround()) {
                         if ((!flag2 || !flag3) && !(this.tracker instanceof EntityArrow)) {
                             if (flag2) {
+                                if (!(k == 0 && l == 0 && i1 == 0)) // Hydrinity - filter useless entity packets
                                 packet1 = new PacketPlayOutEntity.PacketPlayOutRelEntityMove(this.tracker.getId(), (short) ((int) k), (short) ((int) l), (short) ((int) i1), this.tracker.isOnGround());
                             } else if (flag3) {
+                                if (!(i == 0 && j == 0)) // Hydrinity - filter useless entity packets
                                 packet1 = new PacketPlayOutEntity.PacketPlayOutEntityLook(this.tracker.getId(), (byte) i, (byte) j, this.tracker.isOnGround());
                             }
                         } else {
+                            if (!(k == 0 && l == 0 && i1 == 0 && i == 0 && j == 0)) // Hydrinity - filter useless entity packets
                             packet1 = new PacketPlayOutEntity.PacketPlayOutRelEntityMoveLook(this.tracker.getId(), (short) ((int) k), (short) ((int) l), (short) ((int) i1), (byte) i, (byte) j, this.tracker.isOnGround());
                         }
                     } else {
